src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,689 @@
+        uint256 outstandingWithdrawals = withdrawsQueued - withdrawsClaimed;
+
+        // The amount of the liquidity asset is in the ARM
+        uint256 liquidityBalance = IERC20(liquidityAsset).balanceOf(address(this));
+
+        // If there is not enough liquidity assets in the ARM to cover the outstanding withdrawals
+        if (liquidityBalance <= outstandingWithdrawals) {
+            return 0;
+        }
+
+        return liquidityBalance - outstandingWithdrawals;
+    }
+
+    /// @notice The total amount of assets in the ARM and external withdrawal queue,
+    /// less the liquidity assets reserved for the ARM's withdrawal queue and accrued fees.
+    function totalAssets() public view virtual returns (uint256) {
+        uint256 totalAssetsBeforeFees = _rawTotalAssets();
+
+        // If the total assets have decreased, then we don't charge a performance fee
+        if (totalAssetsBeforeFees <= lastTotalAssets) return totalAssetsBeforeFees;
+
+        // Calculate the increase in assets since the last time fees were calculated
+        uint256 assetIncrease = totalAssetsBeforeFees - lastTotalAssets;
+
+        // Calculate the performance fee and remove from the total assets before new fees are removed
+        return totalAssetsBeforeFees - ((assetIncrease * fee) / FEE_SCALE);
+    }
+
+    /// @dev Calculate the total assets in the ARM, external withdrawal queue,
+    /// less liquidity assets reserved for the ARM's withdrawal queue and past accrued fees.
+    /// The accrued fees are from the last time fees were calculated.
+    function _rawTotalAssets() internal view returns (uint256) {
+        // Get the assets in the ARM and external withdrawal queue
+        uint256 assets = token0.balanceOf(address(this)) + token1.balanceOf(address(this)) + _externalWithdrawQueue();
+
+        // Load the queue metadata from storage into memory
+        uint256 queuedMem = withdrawsQueued;
+        uint256 claimedMem = withdrawsClaimed;
+
+        // If the ARM becomes insolvent enough that the total value in the ARM and external withdrawal queue
+        // is less than the outstanding withdrawals.
+        if (assets + claimedMem < queuedMem) {
+            return 0;
+        }
+
+        // Need to remove the liquidity assets that have been reserved for the withdrawal queue
+        // and any accrued fees
+        return assets + claimedMem - queuedMem - feesAccrued;
+    }
+
+    /// @dev Hook for calculating the amount of assets in an external withdrawal queue like Lido or OETH
+    /// This is not the ARM's withdrawal queue
+    function _externalWithdrawQueue() internal view virtual returns (uint256 assets);
+
+    /// @notice Calculates the amount of shares for a given amount of liquidity assets
+    function convertToShares(uint256 assets) public view returns (uint256 shares) {
+        uint256 totalAssetsMem = totalAssets();
+        shares = (totalAssetsMem == 0) ? assets : (assets * totalSupply()) / totalAssetsMem;
```

Comment:
🟥 This scares me. Seems like a potential discontinuity that could boost the value of shares. It's also not symmetrical (outside rounding) with `convertToAssets()`. We know that totalAssets is allowed to return a 0. If for any reason totalSupply was greater than totalAssets, this gives the user too many coins.

What if we did a different approach:

- We send 1e12 raw WETH to the contract when initializing the contract.
-  `_rawTotalAssets` requires that it returns >= 1e12 assets, or it reverts.
- We remove this ternary statement, and just go with a clean mul div round down here in `convertToShares`

Reply:
I wonder if it would be better to have an up-only (outside despots/withdraws) user funds storage, rather than the current method of calling everything leftover as user funds.

Reply:
I've been looking at what it takes to get `totalAssets` to zero. It's easy to make the `totalAssets` go down if we allow the prices to cross price 1.0 by 10 basis points. Obviously the larger the asset amount the larger the loses in real terms. But there needs to be some available assets to be swapped in order for the total assets to reduce. That is, assets that are not reserved for in the withdrawal queue or accrued fees.

Using ChatGPT to do the maths and cross checking with some code, it will take 21,290 swaps for the initial totalAssets of 1e12 to get to zero with a lose of 10 basis points on each swap. The larger the initial value the larger the number of swaps required to get to zero totalAssets. The smaller the lose on each swap, the larger the number of swaps.

| Starting assets | lose on each swap | number of swaps |
|---|---|---|
| 1e12 | 0.001 | 21,290 | 
| 1e15 | 0.001 | 28,195 | 
| 1e18 | 0.001 | 35,099 |
| 1e12 | 0.0005 | 41,205 | 
| 1e15 | 0.0005 | 55,017 | 
| 1e18 | 0.0005 | 68,829 |
| 1e12 | 0.0002 | 98,446 | 
| 1e15 | 0.0002 | 132,981 | 
| 1e18 | 0.0002 | 167,516 |
| 1e12 | 0.0001 | 189,970  | 
| 1e15 | 0.0001 | 259,044  | 
| 1e18 | 0.0001 | 328,118  |

Reply:
Having `totalAssets` as up only is interesting and very possible.

Here's what `totalAssets` currently looks like

```Solidity
/// @notice The total amount of assets in the ARM and external withdrawal queue,
/// less the liquidity assets reserved for the ARM's withdrawal queue and accrued fees.
function totalAssets() public view virtual returns (uint256) {
    uint256 availableAssetsBeforeFees = _availableAssets();

    // If the available assets have decreased, then we don't charge a performance fee
    if (availableAssetsBeforeFees <= lastAvailableAssets) return availableAssetsBeforeFees;

    // Calculate the increase in assets since the last time fees were calculated
    uint256 assetIncrease = availableAssetsBeforeFees - lastAvailableAssets;

    // Calculate the performance fee and remove from the available assets before new fees are removed
    return availableAssetsBeforeFees - ((assetIncrease * fee) / FEE_SCALE);
}
```

If we change the second line to the following, then we will use the `lastAvailableAssets` if the amount of available assets has dropped.
```Solidity
if (availableAssetsBeforeFees <= lastAvailableAssets) return lastAvailableAssets;
```

This means `totalAssets` will always be >= `totalSupply` and `totalAssets` >= `MIN_TOTAL_SUPPLY` (1e12). That is, totalAssets can't be zero.

Reply:
The new accounting system is so much better.

totalAssets >= MIN_TOTAL_SUPPLY is very good invariant. Let's just make it a require in this function. Comparison requires are trivially cheap. In the case that it can't go to zero, we definitely should not have an if branch for it to go to zero. So we should remove the ternary.

Reply:
🔴 Still chanting "death" at this this ternary

Reply:
I've killed the ternary in the latest push as totalAssets can't be less than `MIN_TOTAL_SUPPLY` (1e12)